src/ewah.zig

```diff
@@ -0,0 +1,252 @@
+const mem = std.mem;
+
+/// Encode or decode a bitset using Daniel Lemire's EWAH codec.
+/// ("Histogram-Aware Sorting for Enhanced Word-Aligned Compression in Bitmap Indexes")
+///
+/// EWAH uses only two types of words, where the first type is a 64-bit verbatim ("literal") word.
+/// The second type of word is a marker word:
+/// * The first bit indicates which uniform word will follow.
+/// * The next 31 bits are used to store the number of uniform words.
+/// * The last 32 bits are used to store the number of literal words following the uniform words.
+/// EWAH bitmaps begin with a marker word. A 'marker' looks like:
+///
+///     [uniform_bit:u1][uniform_word_count:u31(LE)][literal_word_count:u32(LE)]
+///
+/// and is immediately followed by `literal_word_count` 64-bit literals.
+/// When decoding a marker, the uniform words precede the literal words.
+///
+/// This encoding requires that the architecture is little-endian with 64-bit words.
+pub fn EWAH(comptime Word: type) type {
+    const word_bits = @bitSizeOf(Word);
+
+    const marker_uniform_word_count_max = (1 << ((word_bits / 2) - 1)) - 1;
+    const marker_literal_word_count_max = (1 << (word_bits / 2)) - 1;
+
+    const MarkerUniformCount = std.math.IntFittingRange(0, marker_uniform_word_count_max); // Word=usize → u31
+    const MarkerLiteralCount = std.math.IntFittingRange(0, marker_literal_word_count_max); // Word=usize → u32
+
+    const Marker = packed struct {
+        uniform_bit: u1, // Whether the uniform word is all 0s or all 1s.
+        uniform_word_count: MarkerUniformCount, // 31-bit number of uniform words following the marker.
+        literal_word_count: MarkerLiteralCount, // 32-bit number of literal words following the uniform words.
+    };
+
+    comptime {
+        assert(@import("builtin").target.cpu.arch.endian() == std.builtin.Endian.Little);
+        assert(@typeInfo(Word).Int.signedness == .unsigned);
+        assert(word_bits % 8 == 0); // A multiple of a byte, so that words can be cast to bytes.
+        assert(@bitSizeOf(Marker) == word_bits);
+        assert(@sizeOf(Marker) == @sizeOf(Word));
+
+        assert(@bitSizeOf(MarkerUniformCount) % 2 == 1);
+        assert(std.math.maxInt(MarkerUniformCount) == marker_uniform_word_count_max);
+
+        assert(@bitSizeOf(MarkerLiteralCount) % 2 == 0);
+        assert(std.math.maxInt(MarkerLiteralCount) == marker_literal_word_count_max);
+    }
+
+    return struct {
+        const Self = @This();
+
+        inline fn marker_word(mark: Marker) Word {
+            return @bitCast(Word, mark);
+        }
+
+        /// Decodes the compressed bitset in `source` into `set`. Panics if `source`'s encoding is invalid.
+        /// Returns the number of *words* written to `target_words`.
+        pub fn decode(source: []const u8, target_words: []Word) usize {
+            assert(source.len % @sizeOf(Word) == 0);
+            const source_words = @alignCast(@alignOf(Word), mem.bytesAsSlice(Word, source));
```

Comment:
I think it would make more sense to put the alignment requirements in the function signature to make it more obvious to the caller, same goes for `encode()`.
```suggestion
        pub fn decode(source: []align(@alignOf(Word)) const u8, target_words: []Word) usize {
            assert(source.len % @sizeOf(Word) == 0);
            const source_words = mem.bytesAsSlice(Word, source);
```